File: solutions/java/0124.java
class Solution {
  public int maxPathSum(TreeNode root) {
    maxPathSumDownFrom(root);

    return ans;
  }

  private int ans = Integer.MIN_VALUE;

  private int maxPathSumDownFrom(TreeNode root) {
    if (root == null)
      return 0;

    int left = Math.max(maxPathSumDownFrom(root.left), 0);
    int right = Math.max(maxPathSumDownFrom(root.right), 0);
    ans = Math.max(ans, root.val + left + right);

    return root.val + Math.max(left, right);
  }
}

File: solutions/cpp/0124.cpp
class Solution {
 public:
  int maxPathSum(TreeNode* root) {
    int ans = INT_MIN;

    function<int(TreeNode*)> maxPathSumDownFrom = [&](TreeNode* root) {
      if (!root) return 0;

      const int left = max(maxPathSumDownFrom(root->left), 0);
      const int right = max(maxPathSumDownFrom(root->right), 0);
      ans = max(ans, root->val + left + right);

      return root->val + max(left, right);
    };

    maxPathSumDownFrom(root);

    return ans;
  }
};